core2/pkg/foundation/auth_passwords.go

```diff
@@ -59,6 +59,61 @@ func checkPassword(correctPassword, salt []byte, plainPassword string) bool {
 	return false
 }
 
+func UpdatePassword(tx *db.Transaction, username string, newPassword string, conditionalChange bool, currentPasswordForVerification string) *util.HttpError {
+	if !conditionalChange || currentPasswordForVerification == "" {
+		rows := db.Select[struct {
+			HashedPassword []byte `json:"hashedPassword"`
+			Salt           []byte `json:"salt"`
+		}](
+			tx,
+			`
+ 					select p.hashed_password, p.salt
+                    from auth.principals p
+                    where p.id = :username
+				`,
+			db.Params{
+				"username": username,
+			},
+		)
+		if len(rows) != 1 {
+			return util.HttpErr(http.StatusBadRequest, "Cannot change password for this user")
+		}
+		currentPasswordAndSalt := rows[0]
+		currentPassword := currentPasswordAndSalt.HashedPassword
+		currentSalt := currentPasswordAndSalt.Salt
+
+		if conditionalChange {
+			if currentPasswordForVerification == "" {
+				return util.HttpErr(http.StatusBadRequest, "No password supplied")
+			}
```

Comment:
This can be checked before doing any database transactions and should probably be done at the top as a result.